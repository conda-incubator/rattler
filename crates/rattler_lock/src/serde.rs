use super::{CondaLock, LockMeta, LockedDependency, LockedDependencyKind};
use serde::de::Error;
use serde::{Deserialize, Deserializer, Serialize, Serializer};
use std::cmp::Ordering;
use std::path::PathBuf;

// Version 2: dependencies are now arrays instead of maps
// Version 3: pip has been renamed to pypi
const FILE_VERSION: u32 = 3;

/// A helper struct to deserialize the version field of the lock file and provide potential errors
/// in-line.
#[derive(Serialize)]
#[serde(transparent)]
struct Version(u32);

impl Default for Version {
    fn default() -> Self {
        Self(FILE_VERSION)
    }
}

impl<'de> Deserialize<'de> for Version {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let version = u32::deserialize(deserializer)?;

        if version > FILE_VERSION {
            let binary = if cfg!(test) {
                // When running tests, use a hardcoded name or an environment variable
                String::from(env!("CARGO_PKG_NAME"))
            } else {
                // When not testing, use the current executable's name, e.g. "pixi" for pixi.
                let binary_path =
                    std::env::current_exe().unwrap_or(PathBuf::from(env!("CARGO_PKG_NAME")));
                let binary_file_name = binary_path
                    .file_name()
                    .unwrap_or(env!("CARGO_PKG_NAME").as_ref());
                binary_file_name.to_string_lossy().into_owned()
            };
            return Err(D::Error::custom(format!(
                "found newer lockfile format version {}, but only up to including version {} is supported.\n\
                Please update your {} version.",
                version, FILE_VERSION, binary
            )));
        }

        Ok(Self(version))
    }
}

impl<'de> Deserialize<'de> for CondaLock {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        #[allow(dead_code)]
        #[derive(Deserialize)]
        struct Raw {
            version: Version,
            metadata: LockMeta,
            package: Vec<LockedDependency>,
        }

        let raw = Raw::deserialize(deserializer)?;
        Ok(Self {
            metadata: raw.metadata,
            package: raw.package,
        })
    }
}

impl Serialize for CondaLock {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        #[derive(Serialize)]
        struct Raw<'a> {
            version: Version,
            metadata: &'a LockMeta,
            package: Vec<&'a LockedDependency>,
        }

        // Sort all packages in alphabetical order. We choose to use alphabetic order instead of
        // topological because the alphabetic order will create smaller diffs when packages change
        // or are added.
        // See: https://github.com/conda/conda-lock/issues/491
        let mut sorted_deps = self.package.iter().collect::<Vec<_>>();
        sorted_deps.sort_by(|&a, &b| {
            a.name
                .cmp(&b.name)
                .then_with(|| a.platform.cmp(&b.platform))
                .then_with(|| a.version.cmp(&b.version))
                .then_with(|| match (&a.kind, &b.kind) {
                    (LockedDependencyKind::Conda(a), LockedDependencyKind::Conda(b)) => {
                        a.build.cmp(&b.build)
                    }
                    (LockedDependencyKind::Pypi(_), LockedDependencyKind::Pypi(_)) => {
                        Ordering::Equal
                    }
                    (LockedDependencyKind::Pypi(_), _) => Ordering::Less,
                    (_, LockedDependencyKind::Pypi(_)) => Ordering::Greater,
                })
        });

        let raw = Raw {
            version: Default::default(),
            metadata: &self.metadata,
            package: sorted_deps,
        };

        raw.serialize(serializer)
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use std::path::Path;

    #[test]
    fn read_conda_lock() {
        let err = CondaLock::from_path(
            &Path::new(env!("CARGO_MANIFEST_DIR"))
                .join("../../test-data/conda-lock/forward-compatible-lock.yml"),
        )
        .unwrap_err();

        insta::assert_snapshot!(format!("{}", err), @"found newer lockfile format version 1000, but only up to including version 3 is supported.\nPlease update your rattler_lock version.");
    }
}
