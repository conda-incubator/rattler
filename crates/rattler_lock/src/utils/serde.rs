use chrono::{DateTime, Utc};
use fxhash::FxBuildHasher;
use indexmap::IndexMap;
use pep508_rs::VersionOrUrl;
use rattler_conda_types::MatchSpec;
use rattler_conda_types::{NamelessMatchSpec, PackageName};
use serde::de::Error as _;
use serde::{Deserialize, Deserializer, Serialize, Serializer};
use serde_with::de::DeserializeAsWrap;
use serde_with::ser::SerializeAsWrap;
use serde_with::DisplayFromStr;
use serde_with::{serde_as, DeserializeAs, SerializeAs};
use std::collections::HashSet;
use std::hash::{BuildHasher, Hash};
use std::marker::PhantomData;

pub(crate) struct Timestamp;

impl<'de> DeserializeAs<'de, chrono::DateTime<chrono::Utc>> for Timestamp {
    fn deserialize_as<D>(deserializer: D) -> Result<chrono::DateTime<chrono::Utc>, D::Error>
    where
        D: Deserializer<'de>,
    {
        let timestamp = i64::deserialize(deserializer)?;

        // Convert from milliseconds to seconds
        let microseconds = if timestamp > 253_402_300_799 {
            timestamp * 1_000
        } else {
            timestamp * 1_000_000
        };

        // Convert the timestamp to a UTC timestamp
        Ok(chrono::DateTime::<chrono::Utc>::from_naive_utc_and_offset(
            chrono::NaiveDateTime::from_timestamp_micros(microseconds)
                .ok_or_else(|| D::Error::custom("got invalid timestamp, timestamp out of range"))?,
            chrono::Utc,
        ))
    }
}

impl SerializeAs<chrono::DateTime<chrono::Utc>> for Timestamp {
    fn serialize_as<S>(source: &DateTime<Utc>, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        // Convert the date to a timestamp
        let timestamp: i64 = source.timestamp_millis();

        // Determine the precision of the timestamp.
        let timestamp = if timestamp % 1000 == 0 {
            timestamp / 1000
        } else {
            timestamp
        };

        // Serialize the timestamp
        timestamp.serialize(serializer)
    }
}

/// Used with serde_with to serialize a collection as a sorted collection.
#[derive(Default)]
pub(crate) struct Ordered<T>(PhantomData<T>);

impl<'de, T: Eq + Hash, S: BuildHasher + Default, TAs> DeserializeAs<'de, HashSet<T, S>>
    for Ordered<TAs>
where
    TAs: DeserializeAs<'de, T>,
{
    fn deserialize_as<D>(deserializer: D) -> Result<HashSet<T, S>, D::Error>
    where
        D: Deserializer<'de>,
    {
        let content =
            DeserializeAsWrap::<Vec<T>, Vec<TAs>>::deserialize(deserializer)?.into_inner();
        Ok(HashSet::from_iter(content))
    }
}

impl<T: Ord, HS, TAs: SerializeAs<T>> SerializeAs<HashSet<T, HS>> for Ordered<TAs> {
    fn serialize_as<S>(source: &HashSet<T, HS>, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let mut elements = Vec::from_iter(source.iter());
        elements.sort();
        SerializeAsWrap::<Vec<&T>, Vec<&TAs>>::new(&elements).serialize(serializer)
    }
}

impl<'de, T: Ord, TAs> DeserializeAs<'de, Vec<T>> for Ordered<TAs>
where
    TAs: DeserializeAs<'de, T>,
{
    fn deserialize_as<D>(deserializer: D) -> Result<Vec<T>, D::Error>
    where
        D: Deserializer<'de>,
    {
        let mut content =
            DeserializeAsWrap::<Vec<T>, Vec<TAs>>::deserialize(deserializer)?.into_inner();
        content.sort();
        Ok(content)
    }
}

impl<T: Ord, TAs: SerializeAs<T>> SerializeAs<Vec<T>> for Ordered<TAs> {
    fn serialize_as<S>(source: &Vec<T>, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let mut elements = Vec::from_iter(source.iter());
        elements.sort();
        SerializeAsWrap::<Vec<&T>, Vec<&TAs>>::new(&elements).serialize(serializer)
    }
}

pub(crate) struct MatchSpecMapOrVec;

impl<'de> DeserializeAs<'de, Vec<String>> for MatchSpecMapOrVec {
    fn deserialize_as<D>(deserializer: D) -> Result<Vec<String>, D::Error>
    where
        D: Deserializer<'de>,
    {
        #[serde_as]
        #[derive(Deserialize)]
        #[serde(untagged)]
        enum MapOrVec {
            Vec(Vec<String>),
            Map(
                #[serde_as(as = "IndexMap<_, DisplayFromStr, FxBuildHasher>")]
                IndexMap<PackageName, NamelessMatchSpec, FxBuildHasher>,
            ),
        }

        Ok(match MapOrVec::deserialize(deserializer)? {
            MapOrVec::Vec(v) => v,
            MapOrVec::Map(m) => m
                .into_iter()
                .map(|(name, spec)| MatchSpec::from_nameless(spec, Some(name)).to_string())
                .collect(),
        })
    }
}

pub(crate) struct Pep440MapOrVec;

impl<'de> DeserializeAs<'de, Vec<String>> for Pep440MapOrVec {
    fn deserialize_as<D>(deserializer: D) -> Result<Vec<String>, D::Error>
    where
        D: Deserializer<'de>,
    {
        #[serde_as]
        #[derive(Deserialize)]
        #[serde(untagged)]
        enum MapOrVec {
            Vec(Vec<String>),
            Map(
                #[serde_as(as = "IndexMap<_, DisplayFromStr, FxBuildHasher>")]
                IndexMap<String, pep440_rs::VersionSpecifiers, FxBuildHasher>,
            ),
        }

        Ok(match MapOrVec::deserialize(deserializer)? {
            MapOrVec::Vec(v) => v,
            MapOrVec::Map(m) => m
                .into_iter()
                .map(|(name, spec)| {
                    pep508_rs::Requirement {
                        name,
                        extras: None,
                        version_or_url: if spec.is_empty() {
                            None
                        } else {
                            Some(VersionOrUrl::VersionSpecifier(spec))
                        },
                        marker: None,
                    }
                    .to_string()
                })
                .collect(),
        })
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use serde::Deserialize;
    use serde_with::serde_as;

    #[test]
    fn test_parse_dependencies_as_map_or_vec() {
        #[serde_as]
        #[derive(Deserialize, Eq, PartialEq, Debug)]
        struct Data {
            #[serde_as(deserialize_as = "MatchSpecMapOrVec")]
            dependencies: Vec<String>,
        }

        let data_with_map: Data = serde_yaml::from_str("dependencies:\n  foo: \">3.12\"").unwrap();
        let data_with_vec: Data = serde_yaml::from_str("dependencies:\n- \"foo >3.12\"").unwrap();
        assert_eq!(data_with_map, data_with_vec);
    }
}
