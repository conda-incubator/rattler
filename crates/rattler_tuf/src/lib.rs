use std::collections::BTreeMap;

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
mod serialize;

use ed25519_dalek::{Signature as Ed25519Signature, Verifier, VerifyingKey};
use hex::FromHex;
use sha2::{Digest, Sha256};

#[derive(Debug, Serialize, Deserialize)]
pub struct Signature {
    other_headers: Option<String>,
    signature: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Version(u32);

#[derive(Debug, Serialize, Deserialize)]
pub struct MetadataSpecVersion(String);

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Type {
    Root,
    KeyMgr,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Payload {
    pub delegations: BTreeMap<String, Delegation>,
    pub expiration: DateTime<Utc>,
    pub metadata_spec_version: MetadataSpecVersion,
    pub timestamp: DateTime<Utc>,
    #[serde(rename = "type")]
    pub file_type: Type,
    pub version: Version,
}

impl Payload {
    pub fn hash(&self) -> String {
        let serialized = serialize::canonserialize(&serde_json::to_value(self).unwrap()).unwrap();
        // compute sha256 hash of the serialized payload
        let mut hasher = Sha256::new();

        // Write input message
        hasher.update(serialized);

        // Read hash digest and consume hasher
        let result = hasher.finalize();
        // convert the hash to a hex string
        let hex_string = format!("{:x}", result);

        hex_string
    }
}

#[derive(Debug, Serialize, Deserialize, PartialEq, PartialOrd, Ord, Eq)]
pub struct PublicKey(String);

impl PublicKey {
    pub fn verify(&self, payload: &Payload, signature: &Signature) -> bool {
        let public_key_bytes = <[u8; 32]>::from_hex(&self.0).unwrap();

        let verifying_key = VerifyingKey::from_bytes(&public_key_bytes).unwrap();
        let signature_bytes = hex::decode(&signature.signature).unwrap();
        let ed_signature = Ed25519Signature::try_from(signature_bytes.as_slice()).unwrap();

        let serialized =
            serialize::canonserialize(&serde_json::to_value(payload).unwrap()).unwrap();

        if let Some(other_headers) = &signature.other_headers {
            // in this case we need to hash the payload with the additional header data
            // to make the signature GPG compatible
            let additional_header_data = hex::decode(other_headers).unwrap();

            let mut hasher = Sha256::new();
            hasher.update(&serialized);
            hasher.update(&additional_header_data);
            hasher.update(b"\x04\xff");
            hasher.update((additional_header_data.len() as u32).to_be_bytes());

            let combined_hash = hasher.finalize();

            verifying_key.verify(&combined_hash, &ed_signature).is_ok()
        } else {
            verifying_key.verify(&serialized, &ed_signature).is_ok()
        }
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Delegation {
    pub pubkeys: Vec<PublicKey>,
    pub threshold: u32,
}

#[derive(Debug, Serialize, Deserialize)]
struct Root {
    pub signatures: BTreeMap<PublicKey, Signature>,
    pub signed: Payload,
}

#[cfg(test)]
mod test {
    use std::path::PathBuf;

    fn test_data_dir() -> PathBuf {
        PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("test-data")
    }

    #[test]
    fn test_root() {
        let root_path = test_data_dir().join("demo/1.root.json");
        let root_str = std::fs::read_to_string(root_path).unwrap();
        let root: super::Root = serde_json::from_str(&root_str).unwrap();

        let payload_hash = root.signed.hash();
        assert_eq!(
            payload_hash,
            "8f264e8c9bb38ec36700a2502ef58a39357594ba754089215687344118391c48"
        );

        let (pubkey, sig) = root.signatures.first_key_value().unwrap();
        let verified = pubkey.verify(&root.signed, sig);
        assert!(verified);
    }

    #[test]
    fn test_key_mgr() {
        let keymgr_path = test_data_dir().join("demo/key_mgr.json");
        let keymgr_str = std::fs::read_to_string(keymgr_path).unwrap();
        let keymgr: super::Root = serde_json::from_str(&keymgr_str).unwrap();

        let payload_hash = keymgr.signed.hash();
        assert_eq!(
            payload_hash,
            "fe15c8fcb4e6147ab32a4957db6cceb4a089162d2cb0400e0fbb8f1babeae6ef"
        );

        let (pubkey, sig) = keymgr.signatures.first_key_value().unwrap();
        let verified = pubkey.verify(&keymgr.signed, sig);
        assert!(verified);
    }

    #[test]
    fn test_canonicalize() {
        let root_path = test_data_dir().join("demo/1.root.json");
        let root_str = std::fs::read_to_string(root_path).unwrap();
        let root: super::Root = serde_json::from_str(&root_str).unwrap();

        let serialized =
            super::serialize::canonserialize(&serde_json::to_value(&root.signed).unwrap()).unwrap();
        let serialized_str = std::str::from_utf8(&serialized).unwrap();
        insta::assert_snapshot!(serialized_str);
    }
}
