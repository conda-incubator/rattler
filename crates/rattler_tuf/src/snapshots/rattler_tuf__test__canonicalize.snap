---
source: crates/rattler_tuf/src/lib.rs
expression: serialized_str
---
{
  "delegations": {
    "key_mgr": {
      "pubkeys": [
        "013ddd714962866d12ba5bae273f14d48c89cf0773dee2dbf6d4561e521c83f7"
      ],
      "threshold": 1
    },
    "root": {
      "pubkeys": [
        "c8bd83b3bfc991face417d97b9c0db011b5d256476b602b92fec92849fc2b36c"
      ],
      "threshold": 1
    }
  },
  "expiration": "2021-11-11T20:58:36Z",
  "metadata_spec_version": "0.6.0",
  "timestamp": "2020-11-11T20:58:36Z",
  "type": "root",
  "version": 1
}
